pkg/client/read.go

```diff
@@ -0,0 +1,102 @@
+//
+// Copyright 2025 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	rekornote "github.com/sigstore/rekor-tiles/pkg/note"
+	"github.com/sigstore/sigstore/pkg/signature"
+	"github.com/transparency-dev/formats/log"
+	tclient "github.com/transparency-dev/trillian-tessera/client"
+	"golang.org/x/mod/sumdb/note"
+)
+
+// ReadClient reads checkpoints, tiles, and entry bundles from the tile storage service.
+type ReadClient interface {
+	ReadCheckpoint(context.Context) (*log.Checkpoint, *note.Note, error)
+	ReadTile(context.Context, uint64, uint64, uint8) ([]byte, error)
+	ReadEntryBundle(context.Context, uint64, uint8) ([]byte, error)
+}
+
+type readClient struct {
+	baseURL  *url.URL
+	client   *tclient.HTTPFetcher
+	origin   string
+	verifier note.Verifier
+}
+
+// NewReader creates a new reader client.
+// (This is currently specific to GCS storage).
+func NewReader(readURL, origin string, verifier signature.Verifier, opts ...Option) (ReadClient, error) {
+	cfg := &clientConfig{}
+	for _, o := range opts {
+		o(cfg)
+	}
+	baseURL, err := url.Parse(readURL)
+	if err != nil {
+		return nil, fmt.Errorf("parsing url %s: %w", readURL, err)
+	}
+	noteVerifier, err := rekornote.NewNoteVerifier(origin, verifier)
+	if err != nil {
+		return nil, fmt.Errorf("creating note verifier: %w", err)
+	}
+	httpClient := &http.Client{
+		Transport: createRoundTripper(http.DefaultTransport, cfg.userAgent, url.Values{"alt": []string{"media"}}),
```

Comment:
what's this query argument? is this the GCS specific bit?

Reply:
Yes, without this the request just returns the object metadata instead of the object. I'll add a comment.

Reply:
I see. Is this handled by the load balancer somehow in the actual deployment?

Reply:
Options for an actual deployment are pointing the read client directly to the GCS bucket URL or to a load balancer URL rewriter, but either way the query parameters will need to be passed through to get the raw object.

Reply:
requirement for sigstore clients to inject a GCS specific query parameter seems quite unappealing, no? 

I suppose we can setup a `google_compute_backend_bucket` on the load balancer like https://tuf-repo-cdn.sigstore.dev/ does: something there has to manage this bit since it just works.

Reply:
Thanks for noticing this, there's another endpoint we can use that doesn't need this query parameter https://cloud.google.com/storage/docs/request-endpoints#access_to_public_objects will change it to use that

Reply:
Actually, I think you are mistaken: this works:
https://storage.googleapis.com/tuf-root-staging/11.root.json

(thats https://storage.googleapis.com/<bucket>/<object>)

Reply:
Fixed